lusidtools/cocoon/cocoon_printer.py

```diff
@@ -0,0 +1,171 @@
+    request constituents.
+    :param dict response: response from Lusid-python-tools
+    :param str file_type: file_type of Lusid-python-tools
+    :return: pd.DataFrame: Successful calls from request
+    :return: pd.DataFrame: Error responses from request that fail (APIExceptions: 400 errors)
+    :return: pd.DataFrame: Failed responses that LUSID rejected
+    """
+    items_success, items_failed, items_error = [], [], []
+    if "success" not in response[file_type].keys() or \
+            "errors" not in response[file_type].keys():
+        raise ValueError("Didn't recognise response shape. Must include 'errors', and 'success' keys")
+
+    # get successes and failures
+    for batch in response[file_type]["success"]:
+        [items_success.append(batch.values[key].name) for key in batch.values.keys()]
+        if len(response[file_type]["success"][0].failed) > 0:
+            [items_failed.append(batch.failed[key].name) for key in batch.failed.keys()]
+        else:
+            items_failed = []
+
+    # get API exception
+    if len(response[file_type]["errors"]) > 0:
+        status = [item.status for item in response[file_type]["errors"]]
+        reason = [item.reason for item in response[file_type]["errors"]]
+        items_error = np.array([reason, status]).T.tolist()
+        errors = pd.DataFrame(items_error, columns=["error_items", "status"])
+    else:
+        errors = pd.DataFrame([], columns=["error_items", "status"])
+
+    return (pd.DataFrame(items_success, columns=["successful items"]),
+            errors,
+            pd.DataFrame(items_failed, columns=["failed_items"]))
+
+
+def format_portfolios_response(response: dict, file_type: str = "portfolios") -> (pd.DataFrame, pd.DataFrame):
+    """
+    This function unpacks a response from portfolio requests and returns successful and errored statuses for
+    request constituents.
+    :param dict response: response from Lusid-python-tools
+    :param str file_type: file_type of Lusid-python-tools
+    :return: pd.DataFrame: Successful calls from request
+    :return: pd.DataFrame: Error responses from request that fail (APIExceptions: 400 errors)
+    """
+    items_success, items_error = [], []
+
+    if "success" not in response[file_type].keys() or \
+            "errors" not in response[file_type].keys():
+        raise ValueError("Didn't recognise response shape. Must include 'errors', and 'success' keys")
+
+    # get successes
+    items_success = [batch.id.code for batch in response[file_type]["success"]]
+
+    # get API exception
+    if len(response[file_type]["errors"]) > 0:
+        status = [item.status for item in response[file_type]["errors"]]
+        reason = [item.reason for item in response[file_type]["errors"]]
+        items_error = np.array([reason, status]).T.tolist()
+        errors = pd.DataFrame(items_error, columns=["error_items", "status"])
```

Comment:
Again don't need to check length of list here.

Reply:
This code is also repeated across all 5 functions. Take this out into a separate function too and only leave the portfolio specific work in here.

Reply:
`get_errors_from_response()`